Adding torchtitan sample (#574)
aws-samples/awsome-distributed-training: 3.test_cases/21.torchtitan/README.md

## Key Features

torchtitan offers several advanced capabilities:

- [FSDP2](https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md) with per-parameter sharding
- [FP8 Support](https://discuss.pytorch.org/t/distributed-w-torchtitan-enabling-float8-all-gather-in-fsdp2/209323)
- [Async Tensor Parallelism](https://discuss.pytorch.org/t/distributed-w-torchtitan-introducing-async-tensor-parallelism-in-pytorch/209487) in PyTorch
- [Optimized Checkpointing Efficiency](https://discuss.pytorch.org/t/distributed-w-torchtitan-optimizing-checkpointing-efficiency-with-pytorch-dcp/211250) with PyTorch DCP
- [Zero-Bubble Pipeline Parallelism](https://discuss.pytorch.org/t/distributed-w-torchtitan-training-with-zero-bubble-pipeline-parallelism/214420)
- [Context Parallelism for training long context LLMs](https://discuss.pytorch.org/t/distributed-w-torchtitan-breaking-barriers-training-long-context-llms-with-1m-sequence-length-in-pytorch-using-context-parallel/215082/1) (with 1M Sequence Length)

## Setup Instructions

> 

Reviewer: Prerequisites? I guess we need Slurm cluster + Shard filesystem + x number of NVIDIA Hopper(or later) GPU instances?

### 1. Create torchtitan Conda Environment

On your cluster head node, run the `0.create_conda_env.sh` script:

```bash
bash 0.create_conda_env.sh
```

This script:
- Downloads and installs Miniconda and creates a torchtian conda environment named "pt_torchtitan"
- Clones the torchtitan repository from GitHub 
- Installs all required dependencies including PyTorch nightly build with CUDA support and [torchao library](https://github.com/pytorch/ao) for FP8 support

### 2. Download the Tokenizer

> Use the following command to download the Meta Llama 3 tokenizer:

Reviewer: We want instruction on model access something like
```suggestion
First, create a Hugging Face account to retrieve a [token](https://huggingface.co/settings/tokens.). Log in to your account and create an access token from Hugging Face Tokens. Then apply for Llama3.1 weight access from [Meta-Llama-3.1-70B](https://huggingface.co/meta-llama/Llama-3.1-70B) page.

Use the following command to download the Meta Llama 3 tokenizer:
```

2: 2025-03-04 00:44:44,441 - root - INFO - [36mstep: 1990  [32mloss:  3.4370  [33mmemory: 68.57GiB(86.69%)  [34mtps: 6,785  [35mmfu: 39.73%[39m
2: 2025-03-04 00:44:44,441 - root - INFO - [36mstep: 1990  [32mloss:  3.4370  [33mmemory: 68.57GiB(86.69%)  [34mtps: 6,785  [35mmfu: 39.73%[39m
2: 2025-03-04 00:44:44,441 - root - INFO - [36mstep: 1990  [32mloss:  3.4370  [33mmemory: 68.57GiB(86.69%)  [34mtps: 6,785  [35mmfu: 39.73%[39m
3: 2025-03-04 00:44:44,441 - root - INFO - [36mstep: 1990  [32mloss:  3.4370  [33mmemory: 68.57GiB(86.69%)  [34mtps: 6,785  [35mmfu: 39.73%[39m
2: 2025-03-04 00:44:44,441 - root - INFO - [36mstep: 1990  [32mloss:  3.4370  [33mmemory: 68.57GiB(86.69%)  [34mtps: 6,785  [35mmfu: 39.73%[39m
```

## Performance Optimizations

To apply various optimizations that leverage `torch.compile` and FP8 for improved performance, update these entries in your training config file:

```toml
compile = true
...
...

[float8]
enable_float8_linear = true
enable_fsdp_float8_all_gather = true
precompute_float8_dynamic_scale_for_fsdp = true
```

> Applying these optimizations to the llama3_8b.toml config and running on the 4 x p5.48xlarge instances we observeimproved throughput(**15.92%** improvement) and MFU metrics(**from 39.73% -> 46.06%**) compared to the default configuration:

Reviewer: ```suggestion
Applying these optimizations to the llama3_8b.toml config and running on the 4 x p5.48xlarge instances we observe improved throughput(**15.92%** improvement) and MFU metrics(**from 39.73% -> 46.06%**) compared to the default configuration:
```

> # torchtitan Training on SageMaker HyperPod

Reviewer: ```suggestion
# torchtitan Training on AWS
```
